### ISSUE TITLE
Understanding the `easy_margin` parameter in torch implementation of arcface loss

### ISSUE THREAD

**Issue by ElisonSherton** [deepinsight/insightface#2126]
Hi there,

I was going through the [`arcface_loss`](https://github.com/deepinsight/insightface/blob/master/recognition/arcface_torch/losses.py) implementation in this repo and had one query regarding the same.

In the `forward` method, there is this piece of code

```python
if self.easy_margin:
    final_target_logit = torch.where(
        target_logit > 0, cos_theta_m, target_logit)
else:
    final_target_logit = torch.where(
        target_logit > self.theta, cos_theta_m, target_logit - self.sinmm)
```

We need to set the `target_logit` to `cos(theta_{yi} + margin)` unconditionally right? 

In the [`arcface_paper`](https://arxiv.org/pdf/1801.07698.pdf), equation 3 in section 3.1 defines arcface loss as follows.

![image](https://user-images.githubusercontent.com/45599762/194217164-b767db30-febf-4492-b1a3-d3f1bfb03577.png)

As per this loss function the target logit is set to `cos(theta_{yi} + margin)` scaled appropriately at the end and the rest is as is. Where does the `sin_margin * margin` and the condition `target_logit > self.thera` come from?

If someone could help me understand this section, I would be very grateful to you.

Regards,
Vinayak.

**Comment by ElisonSherton** [comment]
Hey everyone,

The reason for that conditional update lies in the face that after addition of margin to the angle between target center and the feature vector, the target logit wouldn't continue to stay a monotonically decreasing function.

![image](https://user-images.githubusercontent.com/45599762/194738338-d10c0af8-acbd-460b-bb06-9171615f28ac.png)

As seen after theta = 150 degrees, assuming a margin of 30 degrees, the angle would exceed 180 degrees and cosine of angle between 180 degrees - 210 degrees is an increasing function. 

Hence to maintain the decreasing nature of the loss curve, that update helps since it basically says whenever the angle is greater than `pi - margin`, we would manually subtract a fixed quantity i.e. `margin times sin(margin)`. Hence the condition.

The authors have made this clear in the [older version of their paper](https://arxiv.org/pdf/1801.07698v1.pdf) in section 3 under target logit analysis.

Thanks,
Vinayak.

**Comment by vishhvak** [comment]
Thank you so much for the explanation, this makes so much sense. Is there a reason why the fixed quantity was chosen as margin * sin(margin)?

**Comment by ElisonSherton** [comment]
Hi @vishhvak,

Thanks for your kind words. No, I don't exactly know any mathematical justification for the quantity `margin * sin(margin)` However, I did try it out with some values of margin and it did seem to be decent.

i.e. Intuitively you would want to penalize target-fv angles which are more obtuse than `180-margin` higher so we need to add a quantity which would make the logit go below -1 i.e. worst than what cos function can do. This is so that in the optimization, higher angles will be really really forbidden. I tried the quantity out with a few values of margins and it seemed to look fine.

![image](https://user-images.githubusercontent.com/45599762/198239215-af0644e5-2b8c-4f2e-8c75-43ede5f2a51f.png)

Between angles of zero to lower bound, the logit would be made `cos(angle + margin)` but above the lower bound, it would be directly computed as `cos(angle) - m * sin(m)`. 

There is one more thread which discusses this [here](https://github.com/ronghuaiyang/arcface-pytorch/issues/48) and I think I most liked `blizzard's` answer in this thread. 

Hope this helps.

Thanks,
Vinayak.
